core/tauri-macros/src/command.rs

```diff
@@ -40,40 +59,55 @@ pub fn generate_command(attrs: Vec<NestedMeta>, function: ItemFn) -> TokenStream
     ReturnType::Default => false,
   };
 
-  // Split function args into names and types
-  let (mut names, mut types): (Vec<Ident>, Vec<Path>) = function
-    .sig
-    .inputs
-    .iter()
-    .map(|param| {
-      let mut arg_name = None;
-      let mut arg_type = None;
-      if let FnArg::Typed(arg) = param {
-        if let Pat::Ident(ident) = arg.pat.as_ref() {
-          arg_name = Some(ident.ident.clone());
-        }
-        if let Type::Path(path) = arg.ty.as_ref() {
-          arg_type = Some(path.path.clone());
+  let mut invoke_arg_names: Vec<Ident> = Default::default();
+  let mut invoke_arg_types: Vec<Path> = Default::default();
+  let mut call_arguments = Vec::new();
+
+  for (i, param) in function.sig.inputs.clone().into_iter().enumerate() {
+    let mut arg_name = None;
+    let mut arg_type = None;
+    if let FnArg::Typed(arg) = param {
+      if let Pat::Ident(ident) = arg.pat.as_ref() {
+        arg_name = Some(ident.ident.clone());
+      }
+      if let Type::Path(path) = arg.ty.as_ref() {
+        arg_type = Some(path.path.clone());
+      }
+    }
+
+    if i == 0 && with_window {
+      call_arguments.push(quote!(_window));
+      continue;
+    }
+
+    let arg_name_ = arg_name.clone().unwrap();
+    let arg_type =
+      arg_type.unwrap_or_else(|| panic!("Invalid type for arg \"{}\"", arg_name.unwrap()));
+
+    let mut path_as_string = String::new();
+    for segment in &arg_type.segments {
+      path_as_string.push_str(&segment.ident.to_string());
+      path_as_string.push_str("::");
+    }
+
+    if ["State::", "tauri::State::"].contains(&path_as_string.as_str()) {
```

Comment:
This is problematic because any renamed import will fail. This is also the reason why we went with `with_window` instead of trying to detect the window parameter. Since we can't type detect during the macro, there isn't a name-based way to do this. I think the following choices might be viable:
1. add `with_state` and operate the `with`s in a left to right order
2. create a `FromCommand` trait and have parameters implement it `from_command(msg: InvokeMessage<_>) -> Result<Self>`  (this is similar to what rocket does).  We then use type inference to call that trait method on all the parameters being passed to the function
    * because the "data" parameters aren't name anywhere else, this means we would only realistically be able to accept struct that implment it also, and not use something like `name: String`
3. (maybe, i haven't tested this) accept a `#[window]` and `#[state]` attributes in the parameters, something like the following:
    * ```rust
      #[command]
      fn my_command(
        #[window] window: Window<impl Params>,
        #[state] state: State<'_, String>,
        name: String,
        body: String
      ) {
        todo!()
      }
      ```

ill probably discuss it on discord